core/domain/rules_registry_test.py

```diff
@@ -73,3 +73,32 @@ def test_get_html_field_types_to_rule_specs_for_unsaved_state_schema_version(  #
                 rules_registry.Registry
                 .get_html_field_types_to_rule_specs(state_schema_version=10)
             )
+
+    def test_get_html_field_types_to_rule_specs_for_previous_and_current_state_schema_version(  # pylint: disable=line-too-long
+        self
```

Comment:
@kashishbhandula aren't we testing for `current state schema version` in the previous tests? Or is this testing something else?

Reply:
HI @gp201 , previous test case of  `current state schema version`  and `previous state scheme version` perform half coverage each.
previous test cases were not able to cover the whole file in one go, that way there was coverage missing.
But when I combine them the full file is covered in one go.
PTAL!

Reply:
Ah, I see you seem to be covering a branch.
Could you see which branch (if condtiion) you're covering and probably reflect that in the function name?

Note: Split the tests into 2 separate tests.

Reply:
Hi @gp201, If I split the test case into 2 tests then the problem remains the same, `current state schema version` and `previous state scheme version` will show half coverage each.
In the file already `current state schema version` and `previous state scheme version` test case are written in a separate manner, but when you run the file each test case cover half of the coverage. So it shows missing coverage.
So the solution was to write a test case that can cover both scenarios in one go.
So I merge the `current state schema version` and `previous state scheme version` and it covers both scenario in one go and this result in full file coverage

Reply:
Can you elucidate how that is happening? Meaning why do the previous test cases cover only half and why does combining the two result in complete coverage?

Reply:
Hi @gp201 ,
I have explained the whole process here, I hope this clear all your doubts.
original code:
![Screenshot from 2023-01-28 09-49-25](https://user-images.githubusercontent.com/56387644/215241657-999378a3-d10c-4ef7-8ce1-6f2286772bc3.png)
After I added a print statement:
![Screenshot from 2023-01-28 09-59-33](https://user-images.githubusercontent.com/56387644/215241999-318ef66d-f470-4d96-bc51-4dee7202c246.png)
I added a number of this type of print statement between 72 to 108.
Observation:
They all were executing.
Now the question arises if they all are executing then why missing coverage?
Then i try this part:
If I comment out `current state schema version`
![Screenshot from 2023-01-28 09-48-39](https://user-images.githubusercontent.com/56387644/215241667-3ef327b2-329c-4af5-8c9a-8596d8f3e43c.png)
If I comment out `previous state schema version`
![Screenshot from 2023-01-28 09-46-41](https://user-images.githubusercontent.com/56387644/215241686-107f6480-f371-4fde-966e-0460fcdaa09c.png)
So as you can see some extra lines were added in missing coverage.
Conclusion:
Coverage is full and all lines are executing
but they are not executing in one go(check the screenshot for more understanding).
So i merge them and it worked.

Reply:
Hi @gp201 ,
I misunderstood the statement and I apologize for the mistake, I will not commit such mistakes in future
and sorry for wasting your precious time.
I have fixed my mistakes so PTAL!.

Reply:
Previously, `if not cached:` is never tested in a false manner. https://github.com/oppia/oppia/blob/2286af7b1a29c11feb47e2f5998214760f717219/core/domain/rules_registry.py#L72
I mean till now in every test case `cached` is always false and this is because we are always checking the cases where `cls._state_schema_version_to_html_field_types_to_rule_specs` https://github.com/oppia/oppia/blob/2286af7b1a29c11feb47e2f5998214760f717219/core/domain/rules_registry.py#L70
is never populated before executing `get_html_field_types_to_rule_specs` https://github.com/oppia/oppia/blob/2286af7b1a29c11feb47e2f5998214760f717219/core/domain/rules_registry.py#L47

What I did to accomplish the full coverage? 
I also checked the case where `cls._state_schema_version_to_html_field_types_to_rule_specs` is already populated and `get_html_field_types_to_rule_specs` do not call all the functionality to re-populate the `cls._state_schema_version_to_html_field_types_to_rule_specs` because `cached` in this case is true.

let me know if I missed something!

Reply:
@kashishbhandula thanks for the explanation!! Can you keep them as separate tests but mention the caching in the function name?

Reply:
Hi @gp201, I have added a separate test case with caching in the function name.